File: 0402-remove-k-digits/0402-remove-k-digits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function(num, k) {
    const n = num.length;
    let stack = [];
    for(let ch of num) {
        if(stack.length == 0) {
            stack.push(ch);
            continue;
        }
        while(stack.length > 0 && k > 0 && ch < stack[stack.length - 1]) {
            stack.pop();
            k--;
        }
        stack.push(ch);
    }
    
    while(k > 0 && stack.length > 0) {
        k--;
        stack.pop();
    }
    
    stack = stack.reverse()
    
    while(stack.length > 0 && stack[stack.length - 1] == 0) {
        stack.pop();
    }
    stack = stack.reverse();
    return stack.length <= 0 ? "0" : stack.join('');
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        final int n = num.length();
        Stack<Character> stack = new Stack<Character>();
        for(int i = 0; i < n; i++) {
            if(stack.size() == 0) {
                stack.push(num.charAt(i));
            } else {
                while(stack.size() > 0 && num.charAt(i) < stack.peek() && k > 0) {
                    stack.pop();
                    k--;
                }
                stack.push(num.charAt(i));
            }
        }
        while(stack.size() > 0 && k > 0) {
            stack.pop();
            k--;
        }
        
        Stack<Character> resultStack = new Stack<Character>();
        while(stack.size() > 0) {
            resultStack.push(stack.pop());
        }
        while(resultStack.size() > 0 && resultStack.peek() == '0') {
            resultStack.pop();
        }
        while(resultStack.size() > 0) {
            
            stack.push(resultStack.pop());
        }
        StringBuilder sb = new StringBuilder();
        while(stack.size() > 0) {
            sb.insert(0,stack.pop());
        }
        return sb.length() <= 0 ? "0" : sb.toString();
    }
}